File: host/obc-vectors/src/weather_request.rs
//! The **Weather Request** fixtures (`obc-ble-interface-spec.md` §11, WX3 #1188).
//!
//! Three of them are the `weatherRequestContext` value the companion reads before it disconnects
//! again — the one thing the device says about where the rider is and what forecast it already
//! holds. The other two carry the contract's two *append-only* wire changes, and they belong here
//! rather than beside the older identity/Config fixtures because they only mean anything together:
//! a phone that cannot see the capability word never looks for the service, and a phone that cannot
//! write the refresh byte cannot change how often the device asks.
//!
//! Everything below is written from the spec's field table by explicit offset writes. `obc-vectors`
//! does not depend on `obc-ble`, and must not: a vector produced by the very encoder it exists to
//! police would agree with that encoder by construction and pin nothing at all.

// ============================ The layout (spec §11) ============================

/// The context layout version — v1 is the first, so there is no older form to be lenient towards.
pub const CONTEXT_VERSION: u8 = 1;
/// v1's exact encoded length, and the value byte 1 declares.
pub const CONTEXT_ENCODED_LEN: usize = 52;

// Field offsets, straight off the spec's table. Little-endian throughout.
const OFF_VERSION: usize = 0;
const OFF_ENCODED_LEN: usize = 1;
const OFF_VALIDITY: usize = 2;
const OFF_REASON: usize = 4;
const OFF_REFRESH: usize = 6;
const OFF_RESERVED0: usize = 7;
const OFF_REQUEST_ID: usize = 8;
const OFF_LAT: usize = 12;
const OFF_LON: usize = 16;
const OFF_FIX_UTC: usize = 20;
const OFF_BEARING: usize = 28;
const OFF_SPEED: usize = 30;
const OFF_ROUTE_ID: usize = 32;
const OFF_RESERVED1: usize = 34;
const OFF_BUNDLE_GENERATION: usize = 36;
const OFF_BUNDLE_GENERATED_AT: usize = 40;
const OFF_BUNDLE_CRC32: usize = 48;

/// `lat_udeg` / `lon_udeg` / `fix_utc` carry a real fix.
pub const VALID_POSITION: u16 = 1 << 0;
/// `bearing_deg` is a bearing the device trusts.
pub const VALID_BEARING: u16 = 1 << 1;
/// `speed_deci_ms` is a ground speed the device trusts.
pub const VALID_SPEED: u16 = 1 << 2;
/// The three `bundle_*` fields describe a bundle the device holds and has validated.
pub const VALID_BUNDLE: u16 = 1 << 3;
/// `route_id` names the active route object.
pub const VALID_ROUTE: u16 = 1 << 4;

/// The configured interval elapsed mid-ride.
pub const REASON_SCHEDULED: u16 = 1 << 0;
/// The rider opened Weather — fetch now.
pub const REASON_URGENT: u16 = 1 << 1;
/// A step on the retry ladder.
pub const REASON_RETRY: u16 = 1 << 2;
/// Nothing usable on the card (or the held bundle expired).
pub const REASON_NO_BUNDLE: u16 = 1 << 3;
/// The rider left the held bundle's covered corridor.
pub const REASON_OUT_OF_AREA: u16 = 1 << 4;

/// Scheduled refresh **off** — a device that never raises a *scheduled* request, which is not the
/// same as a device that never raises one (see [`no_fix`]).
pub const REFRESH_OFF: u8 = 0;
/// Every 30 minutes — the device default (epic #1185), and what an absent Config field means.
pub const REFRESH_EVERY_30: u8 = 2;
/// Every 60 minutes.
pub const REFRESH_EVERY_60: u8 = 3;

/// The capability bit that says this device implements the whole §11 contract (`feature_bits` in
/// the identity read). One bit covers the service, the context, object type 20 and the Config
/// field, because a phone that has only some of those has nothing it can do.
pub const FEATURE_WEATHER: u32 = 1 << 0;

// ============================ The fixtures' pinned values ============================

/// Freiburg im Breisgau, in the microdegrees the OBCW header uses — inside the OBCW fixtures'
/// bounding box, so the rider in `weather-request-context-full.bin` is genuinely under the forecast
/// the same fixture says they hold. Transposed lat/lon lands in Somalia and is noticed.
pub const FULL_LAT_UDEG: i32 = 47_999_008;
pub const FULL_LON_UDEG: i32 = 7_842_104;
/// The full context's nonce. The high half names the issue that allocated the layout so a value
/// spotted in a log is traceable; the low half is the per-boot counter.
pub const FULL_REQUEST_ID: u32 = 0x1188_0001;
/// The no-fix context's nonce — a *different* request, so a decoder that reads the wrong file is
/// visible rather than merely wrong about the flags.
pub const NO_FIX_REQUEST_ID: u32 = 0x1188_0002;
/// Heading north-north-west at 7.1 m/s (≈ 25.6 km/h): plausible for a loaded touring bike, and
/// neither value is symmetric enough to survive a byte swap unnoticed.
pub const FULL_BEARING_DEG: u16 = 342;
pub const FULL_SPEED_DECI_MS: u16 = 71;
/// The active route is id 7 — the same waypoint route `route-list.bin` catalogs and
/// `status-download-announce.bin` announces, so "the rider is on a route" is a claim the rest of
/// the fixture set can actually back up.
pub const FULL_ROUTE_ID: u16 = 7;
/// The OBCW fixture the full context says the device already holds. The DWD-shaped one, because
/// that is the realistic thing to be riding on: a 96 × 96 raster of native 15-minute frames.
pub const HELD_BUNDLE_FIXTURE: &str = "weather-dwd-96x96-9f.obcw";
/// The store epoch in `version-read-features.bin`, deliberately unlike the `0xA1B2C3D4` the three
/// older identity-read fixtures share: a test that reads the wrong file fails on the epoch rather
/// than silently passing everything except the capability word.
pub const FEATURES_STORE_EPOCH: u32 = 0xC0DE_F00D;
/// The device name in `config-weather-refresh.bin` — different from `config-v1.bin`'s "OBC Tourer"
/// for the same reason, and a different length, so the trailing byte's offset actually moves.
pub const CONFIG_NAME: &str = "OBC Alpine";
/// …and imperial units, so the refresh byte is pinned as sitting *after* a nonzero `units` rather
/// than after a zero that a misaligned reader could mistake for padding.
pub const CONFIG_UNITS: u8 = 1;

// ============================ Builders ============================

/// One context's semantic fields, named after the layout they serialize into. A struct rather than
/// sixteen positional arguments — at this width a positional call is unreviewable, and the whole
/// value of the fixture is that a reader can check it against the spec table by eye.
pub struct ContextFields {
    pub version: u8,
    pub validity: u16,
    pub reason: u16,
    pub refresh: u8,
    pub request_id: u32,
    pub lat_udeg: i32,
    pub lon_udeg: i32,
    pub fix_utc: i64,
    pub bearing_deg: u16,
    pub speed_deci_ms: u16,
    pub route_id: u16,
    pub bundle_generation: u32,
    pub bundle_generated_at: i64,
    pub bundle_crc32: u32,
}

impl ContextFields {
    /// A well-formed v1 value claiming nothing: the resting state of the GATT attribute, which is
    /// what a peer that reads it out of turn must get instead of the last ride's coordinates.
    pub const RESTING: Self = Self {
        version: CONTEXT_VERSION,
        validity: 0,
        reason: 0,
        refresh: REFRESH_EVERY_30,
        request_id: 0,
        lat_udeg: 0,
        lon_udeg: 0,
        fix_utc: 0,
        bearing_deg: 0,
        speed_deci_ms: 0,
        route_id: 0,
        bundle_generation: 0,
        bundle_generated_at: 0,
        bundle_crc32: 0,
    };
}

/// Serialize a context from the spec's field table — every field written at its stated offset, both
/// reserved fields written zero.
pub fn context(f: &ContextFields) -> Vec<u8> {
    let mut b = vec![0u8; CONTEXT_ENCODED_LEN];
    b[OFF_VERSION] = f.version;
    b[OFF_ENCODED_LEN] = CONTEXT_ENCODED_LEN as u8;
    b[OFF_VALIDITY..OFF_VALIDITY + 2].copy_from_slice(&f.validity.to_le_bytes());
    b[OFF_REASON..OFF_REASON + 2].copy_from_slice(&f.reason.to_le_bytes());
    b[OFF_REFRESH] = f.refresh;
    b[OFF_RESERVED0] = 0;
    b[OFF_REQUEST_ID..OFF_REQUEST_ID + 4].copy_from_slice(&f.request_id.to_le_bytes());
    b[OFF_LAT..OFF_LAT + 4].copy_from_slice(&f.lat_udeg.to_le_bytes());
    b[OFF_LON..OFF_LON + 4].copy_from_slice(&f.lon_udeg.to_le_bytes());
    b[OFF_FIX_UTC..OFF_FIX_UTC + 8].copy_from_slice(&f.fix_utc.to_le_bytes());
    b[OFF_BEARING..OFF_BEARING + 2].copy_from_slice(&f.bearing_deg.to_le_bytes());
    b[OFF_SPEED..OFF_SPEED + 2].copy_from_slice(&f.speed_deci_ms.to_le_bytes());
    b[OFF_ROUTE_ID..OFF_ROUTE_ID + 2].copy_from_slice(&f.route_id.to_le_bytes());
    b[OFF_RESERVED1..OFF_RESERVED1 + 2].copy_from_slice(&0u16.to_le_bytes());
    b[OFF_BUNDLE_GENERATION..OFF_BUNDLE_GENERATION + 4].copy_from_slice(&f.bundle_generation.to_le_bytes());
    b[OFF_BUNDLE_GENERATED_AT..OFF_BUNDLE_GENERATED_AT + 8].copy_from_slice(&f.bundle_generated_at.to_le_bytes());
    b[OFF_BUNDLE_CRC32..OFF_BUNDLE_CRC32 + 4].copy_from_slice(&f.bundle_crc32.to_le_bytes());
    b
}

/// `(generation, generated_at, whole-object CRC-32)` of the OBCW bundle the full context says the
/// device holds — **read out of the checked-in bundle fixture** rather than written here as three
/// literals. The point of the `bundle_*` group is that the phone can tell whether the forecast on
/// the card is the one it would have produced, so a fixture whose bundle identity named no real
/// bundle would exercise the field widths and nothing else.
fn held_bundle() -> (u32, i64, u32) {
    use obc_formats::obcw::{HDR_GENERATED_AT, HDR_GENERATION};
    let bytes = super::obcw::dwd_shaped();
    let generation = u32::from_le_bytes(bytes[HDR_GENERATION..HDR_GENERATION + 4].try_into().unwrap());
    let generated_at = i64::from_le_bytes(bytes[HDR_GENERATED_AT..HDR_GENERATED_AT + 8].try_into().unwrap());
    (generation, generated_at, super::crc32(&bytes))
}

/// The **full** context: a rider mid-ride with everything the device can know — a fix, a bearing, a
/// speed, an active route and a validated bundle — asking for the refresh its own 30-minute
/// schedule just came due for.
///
/// The timing is arithmetic rather than decoration: `fix_utc` is exactly 30 minutes after the held
/// bundle's `generated_at`, so `REASON_SCHEDULED` with `REFRESH_EVERY_30` is a claim the numbers in
/// the same file support. Every validity bit is set, which is what makes this the fixture a second
/// implementation is held to: no field is left at zero where a wrong offset could hide.
pub fn full() -> Vec<u8> {
    let (bundle_generation, bundle_generated_at, bundle_crc32) = held_bundle();
    context(&ContextFields {
        validity: VALID_POSITION | VALID_BEARING | VALID_SPEED | VALID_BUNDLE | VALID_ROUTE,
        reason: REASON_SCHEDULED,
        refresh: REFRESH_EVERY_30,
        request_id: FULL_REQUEST_ID,
        lat_udeg: FULL_LAT_UDEG,
        lon_udeg: FULL_LON_UDEG,
        fix_utc: bundle_generated_at + 30 * 60,
        bearing_deg: FULL_BEARING_DEG,
        speed_deci_ms: FULL_SPEED_DECI_MS,
        route_id: FULL_ROUTE_ID,
        bundle_generation,
        bundle_generated_at,
        bundle_crc32,
        ..ContextFields::RESTING
    })
}

/// The **resting** context: nothing is due, nothing is claimed.
///
/// This is the value the characteristic holds between requests, and it exists as a fixture because
/// it is the one every implementation is tempted to skip — an attribute left at all-zeroes would
/// decode as version 0 and refresh `Off`, i.e. as a device that both speaks an unknown layout and
/// has weather switched off, neither of which is true.
pub fn empty() -> Vec<u8> {
    context(&ContextFields::RESTING)
}

/// An **urgent request with nothing behind it**: the rider opened Weather before the receiver has a
/// fix and with no bundle on the card. No validity bits, a nonzero `request_id`, and — the part
/// worth pinning — `refresh` **Off**.
///
/// A request is still a request when the schedule is off: `REFRESH_OFF` configures the *scheduled*
/// interval, not whether the device may ever ask. A phone that gated on the refresh byte would
/// answer every rider except the ones who turned automatic refresh off and then asked by hand.
pub fn no_fix() -> Vec<u8> {
    context(&ContextFields {
        reason: REASON_URGENT | REASON_NO_BUNDLE,
        refresh: REFRESH_OFF,
        request_id: NO_FIX_REQUEST_ID,
        ..ContextFields::RESTING
    })
}

/// The **11-byte** `protocolVersion` read (spec §1): the 7-byte read every current device serves,
/// plus the trailing `feature_bits u32`.
///
/// Composed from [`super::version_read`] on purpose — the capability word is an *append*, and a
/// builder that rewrote the whole layout to add it would stop demonstrating that.
pub fn version_read_features(version: u16, store_epoch: u32, obcm_version: u8, feature_bits: u32) -> Vec<u8> {
    let mut v = super::version_read(version, store_epoch, obcm_version);
    v.extend_from_slice(&feature_bits.to_le_bytes());
    v
}

/// A Config blob carrying the trailing refresh byte (spec §7.3): `name_len u16 · name · units u8 ·
/// weather_refresh u8`.
///
/// The pair with `config-v1.bin` is the fixture: the same object with and without the appended
/// field, so a decoder is pinned on both "an old app's write leaves refresh unspecified" and "a new
/// app's write is read at the right offset", which is a single off-by-one apart.
pub fn config_weather_refresh(name: &str, units: u8, refresh: u8) -> Vec<u8> {
    let name = name.as_bytes();
    let mut v = Vec::with_capacity(2 + name.len() + 2);
    v.extend_from_slice(&(name.len() as u16).to_le_bytes());
    v.extend_from_slice(name);
    v.push(units);
    v.push(refresh);
    v
}

File: companion-ios/Packages/OBCKit/Sources/OBCTransport/BLE/WeatherRequest.swift
import Foundation
import OBCDomain

// The **Weather Request** contract (spec §11, WX3 / #1188) — the Swift mirror of
// `obc_ble::weather_request`, field-for-field. No CoreBluetooth here: this is pure codec plus the
// one-shot's result/error vocabulary, so it stays host-testable. `specs/vectors/weather-request-*`
// pins both halves against the same bytes.
//
// The shape of the exchange, and why it is cheap enough for a phone's background budget: the
// device swaps its *advertised* service UUID to the Weather Request service while a refresh is due;
// iOS wakes on the service match, connects, reads **one** `WeatherRequestContext` — where the rider
// is, where they are heading, what bundle they already hold — and disconnects. BLE is not held
// across the HTTP that follows. The bundle then rides back as `ObjectType.weatherBundle` on the
// ordinary reliable CoC, stamped with `requestID` so the two connections correlate.

/// Which optional groups of a `WeatherRequestContext` carry real values.
///
/// An `OptionSet` rather than named `Bool`s because the raw word must survive a round-trip
/// verbatim: **unknown bits are ignored, not rejected** — they are how a later firmware says
/// something this build was never going to act on, and refusing a whole read over one would strand
/// a rider's forecast on a bit nobody needed.
public struct WeatherRequestValidity: OptionSet, Equatable, Sendable {
    public let rawValue: UInt16
    public init(rawValue: UInt16) { self.rawValue = rawValue }

    /// The latitude / longitude / fix-time group carries a real GPS fix.
    public static let position = WeatherRequestValidity(rawValue: 1 << 0)
    /// The travel bearing is trustworthy (moving, with a course the device believes).
    public static let bearing = WeatherRequestValidity(rawValue: 1 << 1)
    /// The ground speed is trustworthy.
    public static let speed = WeatherRequestValidity(rawValue: 1 << 2)
    /// The bundle generation / timestamp / CRC group describes a bundle the device has validated
    /// and selected. Clear means *no usable bundle on the card* — not "generation 0".
    public static let bundle = WeatherRequestValidity(rawValue: 1 << 3)
    /// The route id names the active route object.
    public static let route = WeatherRequestValidity(rawValue: 1 << 4)
}

/// Why a request is due — advisory scheduling help for the phone, never an upload gate. A phone
/// that recognises none of the bits still performs the full fetch.
public struct WeatherRequestReason: OptionSet, Equatable, Sendable {
    public let rawValue: UInt16
    public init(rawValue: UInt16) { self.rawValue = rawValue }

    /// The configured refresh interval elapsed during an active ride.
    public static let scheduled = WeatherRequestReason(rawValue: 1 << 0)
    /// The rider opened Weather — treat as urgent.
    public static let urgent = WeatherRequestReason(rawValue: 1 << 1)
    /// A previous attempt failed; this is a step on the 5/10/20-minute retry ladder.
    public static let retry = WeatherRequestReason(rawValue: 1 << 2)
    /// There is no usable bundle at all, or the active one has expired.
    public static let noBundle = WeatherRequestReason(rawValue: 1 << 3)
    /// The rider has travelled outside the active bundle's covered corridor.
    public static let outOfArea = WeatherRequestReason(rawValue: 1 << 4)
    /// The held bundle contains hourly data but no rain frames; a manifest probe alone cannot prove
    /// that recreating it would be unchanged.
    public static let hourlyOnly = WeatherRequestReason(rawValue: 1 << 5)
}

/// The one value the companion reads before it disconnects — **52 little-endian bytes** describing
/// the request and the rider (spec §11).
///
/// ```text
///    0  u8   version = 1
///    1  u8   encoded_len = 52          the writer's own declared length
///    2  u16  validity flags
///    4  u16  reason flags
///    6  u8   refresh (WeatherRefresh, raw — unknown values tolerated)
///    7  u8   reserved = 0
///    8  u32  request_id
///   12  i32  lat_udeg                  ┐
///   16  i32  lon_udeg                  ├ .position
///   20  i64  fix_utc                   ┘
///   28  u16  bearing_deg 0…359         .bearing
///   30  u16  speed_deci_ms             .speed
///   32  u16  route_id                  .route
///   34  u16  reserved = 0
///   36  u32  bundle_generation         ┐
///   40  i64  bundle_generated_at       ├ .bundle
///   48  u32  bundle_crc32              ┘
/// ```
///
/// Optional groups are guarded by `validity` bits rather than by sentinel values, which is why the
/// group accessors below (`fix`, `bundle`, `bearingDegrees`, …) are the intended read path: reading
/// the flat storage directly would let "no fix" pass as the equator and "no bundle" as generation
/// 0. Widths mirror the OBCW header exactly (`i32` microdegrees, `i64` UTC seconds, `u32`
/// generation/CRC) so a value round-trips into a bundle header without narrowing.
public struct WeatherRequestContext: Equatable, Sendable {
    /// The layout version this build writes. v1 is exactly `encodedLength` bytes.
    public static let currentVersion: UInt8 = 1
    /// v1's exact encoded length, and the value byte 1 carries.
    public static let encodedLength = 52
    /// The shortest read that can be classified at all: the version/length prefix itself.
    public static let minimumEncodedLength = 2

    /// Layout version, reported as the device stated it — never normalised away, so a diagnostic
    /// can say which generation answered.
    public var version: UInt8
    public var validity: WeatherRequestValidity
    public var reason: WeatherRequestReason
    /// The device's configured refresh interval **as the byte arrived**, echoed here so the phone
    /// can schedule its own work without also reading Config.
    ///
    /// Raw for the same reason `validity` and `reason` are raw words: this is a device → phone
    /// read, so a value this build does not know is a *newer firmware*, not a malformed one. Use
    /// `refresh` for the typed view and keep this for a verbatim round-trip (spec §11.8).
    public var refreshRaw: UInt8
    /// The request nonce, stamped into the OBCW header's `request_id`. Monotonic per device boot and
    /// **stable across the retry ladder**, so retries of one request stay one request.
    public var requestID: UInt32

    // The flat wire storage. Each field is always on the wire; only its flag says whether it means
    // anything. Read them through the group accessors below.
    public var latitudeMicrodegrees: Int32
    public var longitudeMicrodegrees: Int32
    public var fixUTCSeconds: Int64
    public var bearingWireDegrees: UInt16
    public var speedDeciMetersPerSecond: UInt16
    public var routeWireID: UInt16
    public var bundleWireGeneration: UInt32
    public var bundleWireGeneratedAtSeconds: Int64
    public var bundleWireCRC32: UInt32

    public init(
        version: UInt8 = WeatherRequestContext.currentVersion,
        validity: WeatherRequestValidity = [],
        reason: WeatherRequestReason = [],
        refreshRaw: UInt8 = WeatherRefresh.deviceDefault.rawValue,
        requestID: UInt32 = 0,
        latitudeMicrodegrees: Int32 = 0,
        longitudeMicrodegrees: Int32 = 0,
        fixUTCSeconds: Int64 = 0,
        bearingWireDegrees: UInt16 = 0,
        speedDeciMetersPerSecond: UInt16 = 0,
        routeWireID: UInt16 = 0,
        bundleWireGeneration: UInt32 = 0,
        bundleWireGeneratedAtSeconds: Int64 = 0,
        bundleWireCRC32: UInt32 = 0
    ) {
        self.version = version
        self.validity = validity
        self.reason = reason
        self.refreshRaw = refreshRaw
        self.requestID = requestID
        self.latitudeMicrodegrees = latitudeMicrodegrees
        self.longitudeMicrodegrees = longitudeMicrodegrees
        self.fixUTCSeconds = fixUTCSeconds
        self.bearingWireDegrees = bearingWireDegrees
        self.speedDeciMetersPerSecond = speedDeciMetersPerSecond
        self.routeWireID = routeWireID
        self.bundleWireGeneration = bundleWireGeneration
        self.bundleWireGeneratedAtSeconds = bundleWireGeneratedAtSeconds
        self.bundleWireCRC32 = bundleWireCRC32
    }

    /// The resting value: a well-formed v1 context with nothing valid and no reason. This is what
    /// the characteristic holds before any request is raised, so a peer that reads it out of turn
    /// gets a structurally valid "nothing is due" rather than stale rider coordinates.
    public static let empty = WeatherRequestContext()

    // MARK: The optional groups

    /// The device's configured refresh interval, or `nil` when it named one this build does not
    /// know — the *read* direction of §11.8.
    ///
    /// `nil` here means **unknown**, not `.off` and not the default: a phone that collapsed it to
    /// either would misreport the rider's own setting back to them. It is not an error, though —
    /// tomorrow's firmware appending a fifth interval must not cost a shipped app its entire
    /// weather path over one byte it was never going to act on.
    public var refresh: WeatherRefresh? { WeatherRefresh(wireByte: refreshRaw) }

    /// Where the rider was, or `nil` when `validity` claims no fix. Absence is a cleared flag, never
    /// a sentinel — without this guard "no fix yet" reads as the Gulf of Guinea.
    public var fix: Fix? {
        guard validity.contains(.position) else { return nil }
        return Fix(
            latitudeMicrodegrees: latitudeMicrodegrees,
            longitudeMicrodegrees: longitudeMicrodegrees,
            utc: Date(timeIntervalSince1970: TimeInterval(fixUTCSeconds))
        )
    }

    /// Travel bearing in whole degrees `0…359`, or `nil` when the device does not vouch for it.
    public var bearingDegrees: UInt16? { validity.contains(.bearing) ? bearingWireDegrees : nil }

    /// Ground speed in metres per second, or `nil` when the device does not vouch for it. The wire
    /// carries tenths.
    public var speedMetersPerSecond: Double? {
        validity.contains(.speed) ? Double(speedDeciMetersPerSecond) / 10 : nil
    }

    /// The active route's object id, or `nil` when no route is being navigated. `nil` rather than
    /// `0`, which is a legal object id.
    public var routeID: UInt16? { validity.contains(.route) ? routeWireID : nil }

    /// The bundle the device already holds, or `nil` when it holds none. `nil` rather than
    /// generation 0 — the phone must be able to tell "has nothing" from "has the first one".
    public var bundle: HeldBundle? {
        guard validity.contains(.bundle) else { return nil }
        return HeldBundle(
            generation: bundleWireGeneration,
            generatedAt: Date(timeIntervalSince1970: TimeInterval(bundleWireGeneratedAtSeconds)),
            crc32: bundleWireCRC32
        )
    }

    /// A GPS fix as the request context carries it.
    public struct Fix: Equatable, Sendable {
        public var latitudeMicrodegrees: Int32
        public var longitudeMicrodegrees: Int32
        public var utc: Date

        public init(latitudeMicrodegrees: Int32, longitudeMicrodegrees: Int32, utc: Date) {
            self.latitudeMicrodegrees = latitudeMicrodegrees
            self.longitudeMicrodegrees = longitudeMicrodegrees
            self.utc = utc
        }

        public var latitude: Double { Double(latitudeMicrodegrees) / 1_000_000 }
        public var longitude: Double { Double(longitudeMicrodegrees) / 1_000_000 }
    }

    /// The weather bundle the device has validated and selected.
    public struct HeldBundle: Equatable, Sendable {
        public var generation: UInt32
        public var generatedAt: Date
        public var crc32: UInt32

        public init(generation: UInt32, generatedAt: Date, crc32: UInt32) {
            self.generation = generation
            self.generatedAt = generatedAt
            self.crc32 = crc32
        }
    }

    // MARK: Codec

    // The one place the offsets are written down, so `encode` and `decode` cannot drift apart in a
    // way only a byte-level test would catch.
    private enum Offset {
        static let version = 0
        static let encodedLength = 1
        static let validity = 2
        static let reason = 4
        static let refresh = 6
        static let reserved0 = 7
        static let requestID = 8
        static let latitude = 12
        static let longitude = 16
        static let fixUTC = 20
        static let bearing = 28
        static let speed = 30
        static let routeID = 32
        static let reserved1 = 34
        static let bundleGeneration = 36
        static let bundleGeneratedAt = 40
        static let bundleCRC32 = 48
    }

    /// Encode the fixed 52-byte v1 value. The device is the only writer on the real link; this
    /// exists so the round-trip and shared-vector tests can pin the layout from this side too.
    public func encode() -> Data {
        var bytes = [UInt8](repeating: 0, count: Self.encodedLength)
        bytes[Offset.version] = version
        bytes[Offset.encodedLength] = UInt8(Self.encodedLength)
        bytes.writeLE(validity.rawValue, at: Offset.validity)
        bytes.writeLE(reason.rawValue, at: Offset.reason)
        bytes[Offset.refresh] = refreshRaw
        bytes[Offset.reserved0] = 0
        bytes.writeLE(requestID, at: Offset.requestID)
        bytes.writeLE(UInt32(bitPattern: latitudeMicrodegrees), at: Offset.latitude)
        bytes.writeLE(UInt32(bitPattern: longitudeMicrodegrees), at: Offset.longitude)
        bytes.writeLE(UInt64(bitPattern: fixUTCSeconds), at: Offset.fixUTC)
        bytes.writeLE(bearingWireDegrees, at: Offset.bearing)
        bytes.writeLE(speedDeciMetersPerSecond, at: Offset.speed)
        bytes.writeLE(routeWireID, at: Offset.routeID)
        bytes.writeLE(UInt16(0), at: Offset.reserved1)
        bytes.writeLE(bundleWireGeneration, at: Offset.bundleGeneration)
        bytes.writeLE(UInt64(bitPattern: bundleWireGeneratedAtSeconds), at: Offset.bundleGeneratedAt)
        bytes.writeLE(bundleWireCRC32, at: Offset.bundleCRC32)
        return Data(bytes)
    }

    /// Decode a context read.
    ///
    /// The read is **length-declared**: byte 1 states how many bytes the writer produced, and a read
    /// that delivered fewer is refused rather than half-decoded. Bytes past this version's 52 are
    /// **ignored**, so a future firmware that appends a field keeps working against a shipped app —
    /// the same append-only rule the identity read and Config live under.
    ///
    /// Reserved bytes, unknown validity/reason bits and an unknown `refresh` byte are all
    /// **ignored, not rejected**. The refresh byte belongs in that list precisely because this is a
    /// device → phone read (§11.8): a value this build cannot name is a newer firmware naming a
    /// newer interval, and failing the read over it would let one ordinary enum append switch
    /// weather off permanently on every already-shipped app. It rides through verbatim and reads as
    /// `nil` from `refresh`.
    public init(decoding data: Data) throws {
        guard data.count >= Self.minimumEncodedLength else { throw WeatherRequestError.truncated }
        let base = data.startIndex
        let declared = Int(data[base + Offset.encodedLength])
        // A writer claiming fewer bytes than v1 defines is not an *old* writer — v1 is the first
        // version — so it is malformed rather than something to decode leniently.
        guard declared >= Self.encodedLength, data.count >= declared else {
            throw WeatherRequestError.invalidLength
        }
        self.init(
            version: data[base + Offset.version],
            validity: WeatherRequestValidity(rawValue: data.readLE(at: base + Offset.validity)),
            reason: WeatherRequestReason(rawValue: data.readLE(at: base + Offset.reason)),
            refreshRaw: data[base + Offset.refresh],
            requestID: data.readLE(at: base + Offset.requestID),
            latitudeMicrodegrees: Int32(bitPattern: data.readLE(at: base + Offset.latitude)),
            longitudeMicrodegrees: Int32(bitPattern: data.readLE(at: base + Offset.longitude)),
            fixUTCSeconds: Int64(bitPattern: data.readLE(at: base + Offset.fixUTC)),
            bearingWireDegrees: data.readLE(at: base + Offset.bearing),
            speedDeciMetersPerSecond: data.readLE(at: base + Offset.speed),
            routeWireID: data.readLE(at: base + Offset.routeID),
            bundleWireGeneration: data.readLE(at: base + Offset.bundleGeneration),
            bundleWireGeneratedAtSeconds: Int64(bitPattern: data.readLE(at: base + Offset.bundleGeneratedAt)),
            bundleWireCRC32: data.readLE(at: base + Offset.bundleCRC32)
        )
    }
}

/// Why one weather-request read did not produce a context — plus the one Config-write refusal
/// §11.8 defines (`unknownRefresh`), which shares this vocabulary because it is the same wire byte.
///
/// The decode cases split what `obc_ble`'s `DescriptorError` folds into one `Truncated`:
/// `.truncated` is "not even the length prefix arrived", `.invalidLength` is "byte 1 and the read
/// disagree". Both are refusals — nothing is half-decoded either way.
public enum WeatherRequestError: Error, Equatable, Sendable {
    /// No peripheral has ever completed an authenticated session, so there is nothing to scan for.
    case noKnownBondedPeripheral
    case bluetoothUnavailable
    case timedOut
    case connectionDropped
    /// Byte 1 declared a length below v1's 52, or more bytes than the read delivered.
    case invalidLength
    /// Fewer than the two prefix bytes arrived.
    case truncated
    /// A refresh byte this build does not know — an interval it cannot honour, not a default.
    ///
    /// The mirror of `obc_ble::descriptor::DescriptorError::UnknownRefresh`, and its own case for
    /// the same reason: it is the one decode failure whose correct handling depends on the
    /// **direction of travel** (§11.8). It is thrown by exactly one caller,
    /// `DeviceConfig.weatherRefreshToApply()` — the phone → device write. Both read directions
    /// (a context read, a Config read) report unknown and carry on, so this case never surfaces
    /// from `WeatherRequestContext(decoding:)`; a build that threw it there would take the strict
    /// rule to a direction that cannot survive it.
    case unknownRefresh(UInt8)
    case readFailed
    case cancelled
}

/// One completed weather-request read, with timing evidence from the bounded
/// discover → connect → read → disconnect attempt.
///
/// The durations use a monotonic clock and are **diagnostics only** (they feed the background
/// discovery spike's logs); the context is the entire wire value.
public struct WeatherRequestRead: Equatable, Sendable {
    public var context: WeatherRequestContext
    public var discoveryLatency: Duration
    public var connectedDuration: Duration
    /// True when the read rode an existing foreground link instead of its own connection — which is
    /// also why that link must not be dropped when the read completes.
    public var reusedForegroundConnection: Bool

    public init(
        context: WeatherRequestContext,
        discoveryLatency: Duration,
        connectedDuration: Duration,
        reusedForegroundConnection: Bool
    ) {
        self.context = context
        self.discoveryLatency = discoveryLatency
        self.connectedDuration = connectedDuration
        self.reusedForegroundConnection = reusedForegroundConnection
    }
}

public enum WeatherRequestEvent: Equatable, Sendable {
    case completed(WeatherRequestRead)
    case failed(WeatherRequestError)
}

// MARK: - Little-endian access

// Local to this independently pinned 52-byte request layout: unlike the common `Data` helpers,
// this writer targets an `[UInt8]` at fixed offsets and also owns the layout's `UInt64` fields.
extension Array where Element == UInt8 {
    fileprivate mutating func writeLE(_ value: UInt16, at index: Int) {
        self[index] = UInt8(value & 0xFF)
        self[index + 1] = UInt8((value >> 8) & 0xFF)
    }

    fileprivate mutating func writeLE(_ value: UInt32, at index: Int) {
        for byte in 0..<4 { self[index + byte] = UInt8((value >> (8 * UInt32(byte))) & 0xFF) }
    }

    fileprivate mutating func writeLE(_ value: UInt64, at index: Int) {
        for byte in 0..<8 { self[index + byte] = UInt8((value >> (8 * UInt64(byte))) & 0xFF) }
    }
}

extension Data {
    fileprivate func readLE(at index: Int) -> UInt16 {
        UInt16(self[index]) | (UInt16(self[index + 1]) << 8)
    }

    fileprivate func readLE(at index: Int) -> UInt32 {
        (0..<4).reduce(into: UInt32(0)) { $0 |= UInt32(self[index + $1]) << (8 * UInt32($1)) }
    }

    fileprivate func readLE(at index: Int) -> UInt64 {
        (0..<8).reduce(into: UInt64(0)) { $0 |= UInt64(self[index + $1]) << (8 * UInt64($1)) }
    }
}
